primitives/types/module.go

```diff
@@ -14,7 +14,7 @@ type Module interface {
 	Functions() map[sc.U8]Call
 	PreDispatch(call Call) (sc.Empty, error)
 	ValidateUnsigned(source TransactionSource, call Call) (ValidTransaction, error)
-	Metadata() (sc.Sequence[MetadataType], MetadataModule)
+	Metadata(mdGenerator *MetadataTypeGenerator) MetadataModule
```

Comment:
no need to modify the interface. mdGenerator can be added the same way as in api/metadata/module.go

mdGenerator is an internal helper for Metadata and it's an implementation detail while interfaces should only define requirements and adding it this way turns it into a dependency, which doesn't sound right - e.g. introducing a module with simple metadata that doesn't require mdGenerator or devs who will need to implement this interface but may use different approach that doesn't utilise mdGenerator

Reply:
This would mean that we will not longer have to instantiate the **MetadataGenerator** in _api/metadata/module.go_ (rather, pass it as reference from the runtime) as the modules are initialized in the _runtime._go file and if we want the generator to be part of the Module struct, we will need to pass the same **MetadataGenerator** reference instance to each of the modules in the moment of their initialization. 
I remember that we discussed it would be a better idea if the runtime does not know about this **MetadataGenerator** and this was in fact the reason I modified the method signature (interface) to receive it from the Metadata API. Is the suggested approach here better given the progress we have with the task or we should leave it as it is for now ? cc @radkomih @failfmi